src/Catalog/PackageCatalogItemCreator.cs

```diff
@@ -0,0 +1,234 @@
+
+        private PackageCatalogItemCreator(
+            HttpClient httpClient,
+            ITelemetryService telemetryService,
+            ILogger logger,
+            IStorage storage)
+        {
+            _httpClient = httpClient;
+            _telemetryService = telemetryService;
+            _logger = logger;
+            _storage = storage as IAzureStorage;
+        }
+
+        public static PackageCatalogItemCreator Create(
+            HttpClient httpClient,
+            ITelemetryService telemetryService,
+            ILogger logger,
+            IStorage storage)
+        {
+            if (httpClient == null)
+            {
+                throw new ArgumentNullException(nameof(httpClient));
+            }
+
+            if (telemetryService == null)
+            {
+                throw new ArgumentNullException(nameof(telemetryService));
+            }
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
+            return new PackageCatalogItemCreator(httpClient, telemetryService, logger, storage);
+        }
+
+        public async Task<PackageCatalogItem> CreateAsync(
+            FeedPackageDetails packageItem,
+            DateTime timestamp,
+            CancellationToken cancellationToken)
+        {
+            if (packageItem == null)
+            {
+                throw new ArgumentNullException(nameof(packageItem));
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            PackageCatalogItem item = null;
+
+            if (_storage != null)
+            {
+                item = await GetPackageViaStorageAsync(packageItem, cancellationToken);
+            }
+
+            if (item == null)
+            {
+                item = await GetPackageViaHttpAsync(packageItem, timestamp, item, cancellationToken);
+            }
+
+            return item;
+        }
+
+        private async Task<PackageCatalogItem> GetPackageViaStorageAsync(FeedPackageDetails packageItem, CancellationToken cancellationToken)
+        {
+            PackageCatalogItem item = null;
+            var packageName = $"{packageItem.PackageId.ToLowerInvariant()}.{packageItem.PackageVersion.ToLowerInvariant()}.nupkg";
+            var blob = _storage.GetCloudBlockBlobReference(packageName);
+
+            if (blob == null)
+            {
+                _telemetryService.TrackMetric(
+                    TelemetryConstants.NonExistentBlob,
+                    metric: 1,
+                    properties: GetProperties(packageItem, blob));
+            }
+            else
+            {
+                await blob.FetchAttributesAsync(cancellationToken);
+
+                string packageHash = null;
+                var etag = blob.ETag;
+
+                var metadata = await blob.GetMetadataAsync(cancellationToken);
+
+                if (metadata.TryGetValue("SHA512", out packageHash))
```

Comment:
Is there any way that we can verify that the package we're downloading is the same package that we just read from the feed?

Is there a package hash in the feed that we can compare this to?

Reply:
That's a good idea, and yes we can read the package hash from the feed and compare it to the package hash stashed in the blob's metadata.  That would provide even stronger consistency guarantees for package reads.  See [this discussion](https://github.com/NuGet/NuGet.Services.Metadata/pull/322#discussion_r201877706).

Reply:
I won't be implementing this suggestion as part of this PR but feel free to open an issue.  This check doesn't exist today, and in general, as part of my V3 work I'm reluctant to make non-critical behavioral changes like this.